Dev: ocf.py: Flesh out python library for resource agents

krig opened the pull request:
This PR adds various helper functions (like `is_probe` and `ocf_exit_reason`) as well as a metadata generator and run loop.

It also adds a document describing how to use the library.

Usage example:

``` python
import ocf

def start_action(argument):
    print("The start action receives the argument as a parameter: {}".format(argument))


def main():
    metadata = ocf.Metadata("example-agent",
                            shortdesc="This is an example agent",
                            longdesc="An example of how to " +
                            "write an agent in Python using the ocf " +
                            "Python library.")
    metadata.parameter("argument",
                       shortdesc="Example argument",
                       longdesc="This argument is just an example.",
                       content_type="string",
                       default="foobar")
    metadata.action("start", timeout=60)
    ocf.run(metadata,
            handlers={
                "start": start_action
            })

if __name__ == "__main__":
    main()
```

oalbrigt commented:
No need to rewrite the indentation, and it just makes it harder to track the changes.

krig commented:
Leaving the indentation as tabs confuses my editor, and it's not according to PEP 8 :(
But I can revert that part.

krig commented:
Alright, removed the indentation and whitespace changes.

oalbrigt commented on heartbeat/ocf.py:
```diff
@@ -135,3 +135,256 @@ def emit(self, record):
+		sys.exit(OCF_SUCCESS)
+
+	check_required_params()
+	for handler, func in handlers.items():
```
I think it'd be smarter to do "if action in handlers:" instead of the for-loop.

krig replied:
yeah, I'll change it.

krig commented:
I updated the commit and also added a `OCF_ACTION` variable set to `$__OCF_ACTION` if set or else if only one command line argument is provided, to that command line argument.

I'm going to dig into how `ocft` works to see if there is anything needed to make python agents compatible with it as well.

oalbrigt commented:
Sounds good.

krig commented:
Everything seems ocft-compatible and OCF-compliant as far as I can tell, yeah.

krig commented:
Yeah, I considered that. I’ll change that.

oalbrigt commented:
Perfect. Thanks.

krig commented:
Thanks!

oalbrigt commented on doc/writing-python-agents.md:
```diff
@@ -0,0 +1,90 @@
+# Resource Agent guide for Python
```
@krig I think we should move this under dev-guides.

krig replied:
Sure, I'll make a PR.

krig commented:
Done: #1242
